src/firefly/js/ui/SuggestBoxInputField.jsx

```diff
@@ -0,0 +1,175 @@
+    constructor(props) {
+        super(props);
+        this.state = {
+            isOpen: false,
+            displayValue: props.value,
+            valid: true,
+            message: ''
+        };
+
+        this.onValueChange = this.onValueChange.bind(this);
+    }
+
+
+    onValueChange(ev) {
+        var displayValue = get(ev, 'target.value');
+        var {valid,message} = this.props.validator(displayValue);
+        this.setState({displayValue, valid, message, isOpen: true});
+        this.props.fireValueChange({ value : displayValue, message, valid});
+    }
+
+
+    render() {
+
+        const {displayValue, valid, message, highlightedIdx, isOpen } = this.state;
+        var {validator, label, labelWidth, tooltip, inline, fireValueChange, getSuggestions, renderSuggestion, valueOnSuggestion} = this.props;
+        const currentSuggestions = getSuggestions(displayValue);
+
+        const changeHighlighted = (newHighlightedIdx) => {
+            if (newHighlightedIdx !== highlightedIdx) { this.setState({highlightedIdx: newHighlightedIdx}); }
+        };
+
+        // used when input changes are triggered by suggest box
+        const changeValue = (highlightedIdx) => {
+            if (!isUndefined(highlightedIdx)) {
+                const currentSuggestion = currentSuggestions[highlightedIdx];
+                const value = valueOnSuggestion ? valueOnSuggestion(displayValue, currentSuggestion) : currentSuggestion;
+                if (value !== displayValue) {
+                    var {valid,message} = validator(value);
+                    this.setState({
+                        isOpen: false,
+                        highlightedIdx: undefined,
+                        displayValue: value,
+                        valid,
+                        message
+                    });
+                    fireValueChange({value, valid, message});
+                } else {
+                    this.setState({isOpen: false, highlightedIdx: undefined});
+                }
+            } else {
+                this.setState({isOpen: false});
+            }
+        };
+
+        const handleKeyPress = (ev) => {
+            switch (ev.keyCode) {
+                case 13: // enter
+                case 9: // tab
+                    isOpen && changeValue(highlightedIdx);
+                    break;
+                case 27: // escape
+                    this.setState({highlightedIdx : undefined, isOpen: false});
+                    break;
+                case 38: // arrow up
+                    isOpen && changeHighlighted(isUndefined(highlightedIdx) ? currentSuggestions.length - 1 : Math.max(0, highlightedIdx - 1));
+                    break;
+                case 40: // arrow down
+                    isOpen && changeHighlighted(isUndefined(highlightedIdx) ? 0 : Math.min(highlightedIdx + 1, currentSuggestions.length - 1));
+                    break;
+                default:
+                    break;
+            }
+        };
+
```

Comment:
Why are you defining `handleKeyPress` and `changeValue` inside the render function? since they both need to call `setState` why not just make them methods?